README
Image-GeoTIFF-Tiled version 0.04
======================================

A Perl interface to libgeotiff for tiled geoTIFF images, with support 
for extracting and iterating arbitrary shapes.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

The links to libtiff and libgeotiff were configured on my machine (Ubuntu
9.04) and probably need to be modified if they're not located along a standard
search path. Windows installations will probably have problems too - I would
appreciate any assistance in improving the Makefile configuration.

In the github repository there are two main class files (Tiled.pm). They are
exact copies except the one in the top-level directory has the Inline VERSION
uncommented. This one is used for installation by Inline. The one in the lib/
directory is used for manual testing (instead of make, make test).

DEPENDENCIES

This module requires these other modules and libraries:

  libtiff
  libgeotiff

  Inline
  Module::Load
  List::MoreUtils

COPYRIGHT AND LICENCE

Copyright (C) 2009 by Blake Willmarth

This program is free software; you can redistribute it and/or
modify it under the terms of either:

* the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or

* the Artistic License version 2.0.  


Changes
Revision history for Perl extension Image::GeoTIFF::Tiled.

0.07  Wed Jul 21 00:00:00 2010
    - Smaller usgs samples
    - Added more tests
    - Iterator null value now undef, not -1 (some rasters can have negative
      values).
    - Moved get_* and extract_* methods out of C and into Perl. Data is unpacked
      with unpack() and a template and extract_grid and filter_shape no longer
      suffer from memory errors (obviously).
    - Fixed bug in Shape.pm get_parts() where given latitude was not
      being transformed to the middle of the pixel. This caused some latitudes
      to erronously return only 1 vertex when ray-casting. An additional clause
      is now in place in filter_shape() to test for this situation.
    - Renamed many methods to make naming more standardized and sensical
    - Updated pod

0.06  Wed Jul 14 02:24:00 2010
    - Now handles floating-point rasters and bps > 8
    - Added get_iterator_tile[s] methods
    - Exposed constrain_boundary method, which now copies its 
      inputs
    - Added lots of new metadata and sanity checks
    - Added statistics test with new test image (dem)
    - Updated pod
    - Not released to CPAN (passed tests but too many lingering issues)

0.05  Mon Jul 12 15:38:00 2010
    - return value of coordinates are lists
    - removed image reference from Iterator.pm to prevent 
      circular references
    - added adjacencies() to Iterator.pm (adjacent values)
    - added bounds() to Tiled.pm (image boundary)
    - soft-coded null value (-1) to $NULL package variable in 
      Iterator.pm
    - added convenince method get_iterator() to Tiled.pm

0.04  Wed Sep 30 00:00:00 2009
    - improved get_x() efficiency by using part sorting
    - fixed pod code (ticket #51743)
    - lazy test plans in test scripts that require optional
      modules to prevent failed tests

0.03  Wed Sep 02 04:27:00 2009
    - fixed bug where x values during ray-casting were being
      sorted lexically instead of numerically
        --changed get_x() in Shape.pm

0.02  Mon Aug 31 04:10:00 2009
    - allow loading of all multi-part ShapeFile types
        --added reset_points() method in Shape.pm
    - cleaned up distribution
        --removed Tiled.pm from lib/ in make

0.01  Tue Aug 18 17:32:36 2009
	- original version; created by h2xs 1.23 with options
		--compat-version=5.6.0 -PAXn Image::GeoTIFF::Tiled



t/test_contains.pl
#!/usr/bin/perl
use strict;
use warnings;

# require this file

sub test_contains {
    my ( $image, $iter, $shp_shape, $proj ) = @_;
    unless ( defined $iter ) {
        fail( 'Iterator undefined' );
        return;
    }
    # Test that each "next" shape pixel is in fact in the shape
    my $fail  = 0;
    my $total = 0;
    my $success;
    while ( defined( my $val = $iter->next ) ) {
        my ( $x, $y ) =
            map { sprintf( "%.6f", $_ ) }
            $image->pix2proj( $iter->current_coord );
        if ( defined $proj ) {
            ( $y, $x ) = $proj->inverse( $x, $y );
        }
        unless (
            $shp_shape->contains_point(
                Geo::ShapeFile::Point->new( X => $x, Y => $y )
            )
            )
        {
            $fail++;
            warn "Failure at ($x,$y)";
        }
        $total++;
    }
    $success = ( $total - $fail ) / $total;
    ok(
        $success >= 0.99,
        "Geo::ShapeFile::Shape contains agrees with iterator ("
            . sprintf( "%.3f%%", $success * 100 )
            . " success rate)"
      );

# Reverse the buffer values and test the -1 values are in fact outside the shape
    my @old_buffer = @{ $iter->buffer };
    my $evil_buffer;
    for my $i ( 0 .. @{ $iter->buffer } - 1 ) {
        for my $j ( 0 .. @{ $iter->buffer->[ $i ] } - 1 ) {
            $evil_buffer->[ $i ][ $j ] =
                $iter->buffer->[ $i ][ $j ] == -1 ? 1 : -1;
        }
    }
#    print Dumper(\@old_buffer), "\n", Dumper($evil_buffer),"\n";
    $iter->reset;
    $iter->{ buffer } = $evil_buffer;
    $fail             = 0;
    $total            = 0;
    while ( defined( my $val = $iter->next ) ) {
        my ( $x, $y ) =
            map { sprintf( "%.6f", $_ ) }
            $image->pix2proj( $iter->current_coord );
        if ( defined $proj ) {
            ( $y, $x ) = $proj->inverse( $x, $y );
        }
        unless (
            !$shp_shape->contains_point(
                Geo::ShapeFile::Point->new( X => $x, Y => $y )
            )
            )
        {
            $fail = 1;
            warn "Failure at ($x,$y)";
        }
        $total++;
    }
    $success = ( $total - $fail ) / $total;
    ok(
        $success >= 0.99,
        "Geo::ShapeFile::Shape contains agrees with iterator ("
            . sprintf( "%.3f%%", $success * 100 )
            . " success rate)"
      );

    $iter->reset;
    $iter->{ buffer } = \@old_buffer;
} ## end sub test_contains

1;


lib/Image/GeoTIFF/Tiled/Shape.pm
package Image::GeoTIFF::Tiled::Shape;
use strict;
use warnings;
use Carp;
use List::MoreUtils qw(minmax);
use Module::Load;

use Image::GeoTIFF::Tiled;
use Image::GeoTIFF::Tiled::ShapePart;

# Needs to be a series of ordered pixel points

# Data structure: 2D array - one row per pixel row
#   - first row ($data->[0]) is at pixel latitude y_min
#   - last row is at pixel latitude y_max
#   - rows should be ordered on longitude (x) (specific method call when using main method)

use vars qw/ $VERSION /;
$VERSION = '0.07';

#================================================================================================#

sub new {
    my ( $class, $opts ) = @_;
    my $self = {};
    bless( $self, $class );

    if ( ref $opts eq 'ARRAY' ) {
        croak "Need x_min, y_min, x_max, y_max elements."
            unless scalar @$opts == 4;
        $self->x_min( $opts->[ 0 ] );
        $self->y_min( $opts->[ 1 ] );
        $self->x_max( $opts->[ 2 ] );
        $self->y_max( $opts->[ 3 ] );
    }
    elsif ( ref $opts eq 'HASH' ) {
        for ( keys %{ $opts } ) {
            $self->x_min( $opts->{ $_ } ) if $_ eq 'x_min';
            $self->y_min( $opts->{ $_ } ) if $_ eq 'y_min';
            $self->x_max( $opts->{ $_ } ) if $_ eq 'x_max';
            $self->y_max( $opts->{ $_ } ) if $_ eq 'y_max';
        }
    }
    croak "Boundary required." unless grep { defined $_ } $self->boundary;
    return $self;
}

sub load_shape {
    my ( $class, $tiff, $proj, $shape ) = @_;
    croak "loading shapes must be called as class invocant"
        if ref $class and $class ne __PACKAGE__;
    croak "Image::GeoTIFF::Tiled required"
        unless defined $tiff
            and ref $tiff
            and $tiff->isa( "Image::GeoTIFF::Tiled" );
    if ( defined $proj ) {
        croak "Geo::Proj4 required as the projection class"
            unless ref $proj and $proj->isa( "Geo::Proj4" );
        load 'Geo::Proj4';
    }
    my $self;
    if ( ref $shape and $shape->isa( 'Geo::ShapeFile::Shape' ) ) {
        load 'Geo::ShapeFile';    # run-time loading
        my $boundary =
            [ $shape->x_min, $shape->y_min, $shape->x_max, $shape->y_max ];
        Image::GeoTIFF::Tiled::Shape->project_boundary( $proj, $boundary )
            if defined $proj;
        $self = Image::GeoTIFF::Tiled::Shape->new(
            [ $tiff->proj2pix_boundary( @$boundary ) ] );
        for my $i ( 1 .. $shape->num_parts ) {
            $self->reset_points;
            for ( $shape->get_part( $i ) ) {
                my ( $x, $y ) =
                    defined $proj
                    ? ( $proj->forward( $_->Y, $_->X ) )
                    : ( $_->X, $_->Y );
                $self->add_point( $tiff->proj2pix( $x, $y ) );
            }
        }
        $self->finish_loading;
    }
    else {
        croak "Cannot load unknown shape";
    }
    return $self;
} ## end sub load_shape

sub project_boundary {
    my ( $class, $proj, $b ) = @_;
    my @px = ( $b->[ 0 ], $b->[ 0 ], $b->[ 2 ], $b->[ 2 ] );
    my @py = ( $b->[ 3 ], $b->[ 1 ], $b->[ 3 ], $b->[ 1 ] );
    for ( 0 .. 3 ) {
        # See Geo::Proj4 documentation - this is correct
        ( $px[ $_ ], $py[ $_ ] ) = $proj->forward( $py[ $_ ], $px[ $_ ] );
    }
    ( $b->[ 0 ], $b->[ 2 ] ) = minmax( @px );
    ( $b->[ 1 ], $b->[ 3 ] ) = minmax( @py );
}

#================================================================================================#

sub _elem {
    my $self = shift;
    my $key  = shift;
    return $self->{ "_$key" } unless @_;
    $self->{ "_$key" } = $_[ 0 ];
}
# Config accessors
sub x_min { &_elem( shift, 'x_min', @_ ); }
sub y_min { &_elem( shift, 'y_min', @_ ); }
sub x_max { &_elem( shift, 'x_max', @_ ); }
sub y_max { &_elem( shift, 'y_max', @_ ); }

sub boundary {
    my ( $self ) = @_;
    return ( $self->x_min, $self->y_min, $self->x_max, $self->y_max );
}

sub corners {
    my ( $self ) = @_;
    return (
        [ $self->x_min, $self->y_min ],
        [ $self->x_min, $self->y_max ],
        [ $self->x_max, $self->y_min ],
        [ $self->x_max, $self->y_max ]
    );
}

#================================================================================================#

sub finish_loading {
    my ( $self ) = @_;
    $self->reset_points;
    $self->sort_parts;
}

sub sort_parts {
    # Sort parts on latitude (upper / highest point)
    my ( $self ) = @_;
    @{ $self->{ _parts } } =
        sort { $a->upper->[ 1 ] <=> $b->upper->[ 1 ] } @{ $self->{ _parts } };
}

sub reset_points {
    my ( $self ) = @_;
    undef $self->{ $_ } for ( qw/ _first_point _last_end _last_start / );
}

sub add_point {
    my ( $self, $x, $y ) = @_;
    my $point      = [ $x, $y ];
    my $last_start = $self->{ _last_start };
    my $last_end   = $self->{ _last_end };

    # First point
    unless ( defined $self->{ _first_point } ) {
        $self->{ _first_point } = $point;
        $self->{ _last_end }    = $point;
        return;
    }

    # TEST IF $LAST_END IS A LOCAL HORIZONTAL VERTEX - IF SO ADD THE POINT AGAIN
    if ( defined $last_start
        and &_local_hvertex( $last_start, $last_end, $point ) )
    {
        push @{ $self->{ _parts } },
            Image::GeoTIFF::Tiled::ShapePart->new( $last_end, $last_end );
    }

    # Add a new part
    push @{ $self->{ _parts } },
        Image::GeoTIFF::Tiled::ShapePart->new( $last_end, $point );

# Last point (if equals first point) - test if it's a local horizontal vertex (only chance)
    if (    $self->{ _first_point }[ 0 ] == $x
        and $self->{ _first_point }[ 1 ] == $y
        and &_local_hvertex( $last_end, $point, $self->get_part( 0 )->end ) )
    {
        push @{ $self->{ _parts } },
            Image::GeoTIFF::Tiled::ShapePart->new( $point, $point );
    }
    $self->{ _last_start } = $last_end;
    $self->{ _last_end }   = $point;
} ## end sub add_point

sub _local_hvertex {
    # Test $p2
    my ( $p1, $p2, $p3 ) = @_;
    return 1
        if ( $p2->[ 1 ] < $p1->[ 1 ] && $p2->[ 1 ] < $p3->[ 1 ] )
        or ( $p2->[ 1 ] > $p1->[ 1 ] && $p2->[ 1 ] > $p3->[ 1 ] );
    0;
}

sub as_array {
    my $self = shift;
    my @a;
    for ( 0 .. $self->num_parts - 1 ) {
        my $p = $self->get_part( $_ );
        push @a, [ $p->start, $p->end ];
    }
    \@a;
}

#================================================================================================#

sub num_parts {
    my $self = shift;
    return (
        defined $self->{ _parts }
        ? scalar @{ $self->{ _parts } }
        : 0
    );
}

sub get_part {
    my ( $self, $i ) = @_;
    $self->{ _parts }[ $i ];
}

sub get_x {
    # Retrieves all x-values|points along integer latitude y
    # - assume parts are pre-sorted on upper latitude
    my ( $self, $y ) = @_;
    $y = int( $y ) + 0.5;       # Center the latitude!
    my @parts;
    for ( 0 .. $self->num_parts - 1 ) {
        my $part    = $self->get_part( $_ );
        my $upper_y = $part->upper->[ 1 ];
        my $lower_y = $part->lower->[ 1 ];
        # print "$y <=> $upper_y - $lower_y\n";
        last if $y < $upper_y;    # Y above remaining parts
                                  # print $part->str,"\n";
        next unless $y >= $upper_y;
        if ( $y <= $lower_y ) {
            # print "Added part.\n";
            push @parts, $part;
        }
    }
    # @parts now has all parts of the shape containing $y
    # print "Parts:\n",join("\n\n",map $_->str,@parts),"\n";
    my @x =
        sort { $a <=> $b }
        grep defined $_,
        map  { $_->get_x( $y ) } @parts;
    # print "(",join (', ',map { sprintf("%.2f",$_) } @x),")\n";

    \@x;
}

# - NOTE: IF TWO EQUAL X VALUES ARE OBTAINED, IT'S A LOCAL HORIZONTAL VERTEX AND SHOULDN'T CHANGE THE
#   INSIDE/OUTSIDE STATE (or should be changed twice - resulting in the same state AFTER the point is checked)

#================================================================================================#
# POD
1;

=head1 NAME

Image::GeoTIFF::Tiled::Shape

=head1 SYNOPSIS

    use Image::GeoTIFF::Tiled;
    
    # Initiate an instance via a class factory method, importing a pre-existing shape object:
    use Geo::ShapeFile;
    my $shp_shape = ... # A Geo::ShapeFile::Shape retrieved from Geo::ShapeFile methods
    my $shape = Image::GeoTIFF::Tiled::Shape->load_shape( $tiff, $proj, $shp_shape );
    
    # OR Create your own:
    my $shape = Image::GeoTIFF::Tiled::Shape->new({
        x_min => ...,
        y_min => ...,
        x_max => ...,
        y_max => ...
    });
    $shape->add_point($x,$y) for ...;
    $shape->finish_loading;
    
    # Initiate the TIFF image object
    my $t = Image::GeoTIFF::Tiled->new( $tiff );
    
    # Get in iterator for the pixels in the shape
    my $iter = $t->get_iterator( $shape );
    
=head1 DESCRIPTION

This class is meant to be used in conjuction with Image::GeoTIFF::Tiled in order to easily iterate over image pixels contained in an arbitrary shape. It does so by linking Image::GeoTIFF::Tiled::ShapePart's, which are essentially lines between two points, with all intermediate points along integer y-values interpolated.

Objects should be instantiated via the C<load_shape> method for pre-defined shape-like objects (L<Geo::ShapeFile::Shape> being an example). Otherwise they must be instantiated with a boundary, and points should be added sequentially.

Other than constructing a useful data structure for getting an iterator, there isn't any other stated purpose for this class.

All coordinates must be in pixels.

=head1 METHODS

=head2 CONSTRUCTOR

=over

=item new( \%boundary )

=item new( \@boundary )

where %boundary has the x_min, y_min, x_max, and y_max keys filled or @boundary has them in that order. Pixel coordinates required. Without defining a boundary the constructor throws an exception.

=back

=head2 FACTORY METHOD

=over

=item load_shape( $tiff, $proj, $shape )

Loads a pre-defined shape object defined by an external class. Currently only loads L<Geo::ShapeFile::Shape> objects. 

L<Geo::Proj4> and L<Image::GeoTIFF::Tiled> objects must be pre-loaded into the class before calling this method, unless the shape is already projected, in which case pass undef as the $proj parameter.

=back

=head2 METHODS

=over

=item x_min x_max y_min y_max

Retrieves the boundary values.

=item boundary

Equivalent to (C<<$shape->x_min>>, C<<$shape->y_min>>, C<<$shape->x_max>>, C<<$shape->y_max>>).

=item corners

Returns a list of four two-element arrayref's containing the upper left, lower left, upper right, and lower right corner coordinates, in that order.

=item num_parts

Returns the number of L<Image::GeoTIFF::Tiled::ShapePart>'s in this shape.

=item get_part($i)

Returns the ith L<Image::GeoTIFF::Tiled::ShapePart> in this shape.

=item as_array

Returns a 2D array reference of [ start, end ] points corresponding to each part of the shape.

=item add_point($x,$y)

Adds the ($x,$y) point to this shape. Only used for making custom shapes.

=item reset_points

Resets internal points metadata. Only used for making custom shapes, after a series of connecting points are made concentric ("parts" in ShapeFile parlance).

=item sort_parts

Sorts the internal parts array on the upper point latitude (called by finish_loading).

=item finish_loading

Call when done making a custom shape.

=item project_boundary($proj,$b)

Class method. Returns the projected boundary using projection (L<Geo::Proj4>) $proj and boundary $b, a 4-element array ref with x_min, y_min, x_max, y_max as values.

=item get_x($y)

Returns a reference to a sorted array containing all x-pixel values along the integer y latitude. This method is used to determine if a given pixel lies inside the shape by implementing a ray-casting algorithm using a state machine (either outside or inside).

Repeated values indicate a local horizontal vertex.

=back

=head1 SEE ALSO

L<Geo::Proj4>, Image::GeoTIFF::Tiled, Image::GeoTIFF::Tiled::Iterator, Image::GeoTIFF::Tiled::ShapePart, L<Geo::ShapeFile>

=head1 COPYRIGHT & LICENSE

Copyright 2010 Blake Willmarth.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or

=item * the Artistic License version 2.0.

=back

=cut



lib/Image/GeoTIFF/Tiled/ShapePart.pm
package Image::GeoTIFF::Tiled::ShapePart;
use strict;
use warnings;
use Carp;

use vars qw/ $VERSION /;
$VERSION = '0.07';

# Parts are lines (start != end) or horizontal vertexes (start = end)

sub new {
    my ( $class, @p ) = @_;
    my $self = {};
    bless( $self, $class );
    my ( $start, $end );

    if ( ref $p[ 0 ] and ref $p[ 0 ] eq 'ARRAY' and @p == 1 ) {
        confess "2-element array required during constructor."
            unless @{ $p[ 0 ] } == 2;
        $start = $p[ 0 ]->[ 0 ];
        $end   = $p[ 0 ]->[ 1 ];
    }
    elsif ( @p == 2 ) {
        $start = $p[ 0 ];
        $end   = $p[ 1 ];
    }
    else {
        confess "Invalid constructor arguments: @p";
    }

    $self->start( $start );
    $self->end( $end );
    croak "Shape part needs a start and end point."
        unless defined $self->start and defined $self->end;

    $self;
}

sub str {
    my $self = shift;
    my ( $start, $end ) = sort { $a->[ 1 ] <=> $b->[ 1 ] } $self->start,
        $self->end;
    # my ( $p0, $p1 ) = sort { $a->[ 1 ] <=> $b->[ 1 ] } $self->{ _points }[ 0 ],
        # $self->{ _points }[ -1 ];
    my @pts = @{$self->{_points}};
    my ( $u, $l ) = ( $self->upper, $self->lower );
    join( "\n",
        sprintf( "Between:   (%.2f,%.2f) and (%.2f,%.2f)", @$start, @$end ),
        join("\n\t", "Points:", 
            map sprintf( "$_: (%.2f,%.2f)",
                        @{$pts[$_]} ), 0..@pts-1
            ),
        # sprintf( "Upper-Lower: (%.2f,%.2f) and (%.2f,%.2f)", @$u,     @$l ) );
        );
}

sub _point {
    my ( $self, $key, $point ) = @_;
    return $self->{ $key } unless defined $point;
    confess "Point must be 2-element arrayref"
        unless ref $point
            and ref $point eq 'ARRAY'
            and scalar @{ $point } == 2;
    carp "WARNING: Resetting $key value of part" if defined $self->{ $key };
    confess "Point values must be defined"
        if grep { not defined $_ } @{ $point };
#    confess "Point contains negative values" if grep { $_ < 0 } @$point;
    # Set the start/end point
    $self->{ $key } = $point;

    my ( $start, $end ) = ( $self->{ start }, $self->{ end } );
    if ( defined $start and defined $end ) {
        # Ensure _upper at a lower latitude
        if ( $start->[ 1 ] > $end->[ 1 ] ) {
            $self->{ _upper } = $end;
            $self->{ _lower } = $start;
        }
        else {
            $self->{ _upper } = $start;
            $self->{ _lower } = $end;
        }
      # Reset and re-interpolate points if starting and ending point are now set
        $self->_reset_points;
    }
}
sub start { shift->_point( 'start', @_ ); }
sub end   { shift->_point( 'end',   @_ ); }
sub upper { return $_[ 0 ]->{ _upper } }
sub lower { return $_[ 0 ]->{ _lower } }

sub _reset_points {
    my ( $self ) = @_;
    my ( $x0, $y0, $x1, $y1 ) =
        ( @{ $self->{ _upper } }[ 0 .. 1 ], @{ $self->{ _lower } }[ 0 .. 1 ] );

    $self->{ _points } = [];

    return if $y0 == $y1;

    # Interpolate interemediate latitudes (given y, solve for x):
    #   - interpolate the middle of the pixel
    my $y0_ =
        $y0 - int( $y0 ) <=
        0.5                # If in the lower half of the first pixel (inclusive)
        ? int( $y0 ) + 0.5 # start interpolating the middle of that pixel
        : int( $y0 ) + 1.5 # start interpolating the middle of the next pixel
        ;
    my $y1_ =
        $y1 - int( $y1 ) <
        0.5                # If in the lower half of the last pixel (exclusive)
        ? int( $y1 ) -
        0.5    # end interpolation in the middle of the previous pixel
        : int( $y1 ) + 0.5    # end interpolating in the middle of that pixel
        ;

    #   x = x0 + (y - y0) * [ (x1 - x0)/(y1 - y0) ]
    my $factor = ( $x1 - $x0 ) / ( $y1 - $y0 );
    my $i = 0;
    for ( my $y = $y0_; $y <= $y1_; $y++ ) {
        $self->{ _points }[ $i++ ] = [ $x0 + ( $y - $y0 ) * $factor, $y ];
    }
    $self->{ _y0_ } = $y0_;
} ## end sub _reset_points

sub get_point {
    # Get the [ x, y ] value of this part, given the integer y (latitude)
    my ( $self, $y ) = @_;
    confess "Require a y-value (latitude) to retrieve points" unless defined $y;
    return unless defined $self->{ _y0_ };
    my $i = int( $y ) + 0.5 - $self->{ _y0_ };
    # my $i = int( $y ) - 0.5 - $self->{ _y0_ };
    if ( $i < 0 ) {
        # carp "Latitude $y smaller than first point: ", $self->{ _y0_ };
        return;
    }
    $self->{ _points }[ $i ];
}

sub get_x {
    my ( $self, $y ) = @_;
    my $p = $self->get_point( $y );
    return unless defined $p;
    # printf "Point: (%.2f,%.2f)\n",$p->[0],$p->[1];
    return $p->[ 0 ];
}

#================================================================================================#
# POD
1;

=head1 NAME

Image::GeoTIFF::Tiled::ShapePart

=head1 DESCRIPTION

This class is used by L<Image::GeoTIFF::Tiled::Shape> to represent a single "part" of a shape (a line between two points), shapes being made up of multiple parts.

Whenever the start or end points are set, the linear interpolation along integer y-values are calculated and stored. Interpolation is done at the middle of the pixel latitude.

=head1 METHODS

=head2 CONSTRUCTOR

=over

=item new($start,$end)

Starting and ending points are required during construction. Points are 2D array references [ $x, $y ].

=back

=head2 ACCESSORS

=over

=item start

Returns, optionally sets, the starting point. Setting causes the intermediate points between the ending point to be re-interpolated and therefore shouldn't be done.

=item end

Returns, optionally sets, the ending point. Setting causes the intermediate points between the starting point to be re-interpolated and therefore shouldn't be done.

=item upper

Returns the start or end point, whichever has the smaller latitude.

=item lower

Returns the start or end point, whichever has the larger latitude.

=item str

Returns a string representation of the object (debugging only).

=back

=head2 GET POINTS

=over

=item get_point($y)

Returns the [ $x, $y ] interpolated point located at (integer) pixel latitude $y, or C<undef> if there's no point along the given $y.

=item get_x($y)

Returns just the $x value of C<get_point>, or C<undef> if there's no point along the given $y.

=back

=head1 COPYRIGHT & LICENSE

Copyright 2010 Blake Willmarth.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or

=item * the Artistic License version 2.0.

=back

=cut


lib/Image/GeoTIFF/Tiled/Iterator.pm
package Image::GeoTIFF::Tiled::Iterator;
use strict;
use warnings;
use Carp;

use Image::GeoTIFF::Tiled;

use vars qw/ $VERSION /;
$VERSION = '0.07';
# $NULL = undef;
# $NULL    = -1;

#================================================================================================#
# Constructor

sub new {
    my ( $class, $opts ) = @_;
    my $self = {};
    bless( $self, $class );

    $self->reset;

    # $self->image(delete $opts->{image});
    $self->boundary( delete $opts->{ boundary } );
    $self->{ buffer } = delete $opts->{ buffer };
    confess "Invalid options: %$opts" if %$opts;

    my $boundary = $self->boundary;
    $self->rows( int( $boundary->[ 3 ] ) - int( $boundary->[ 1 ] ) + 1 )
        ;    # pixel rows
    $self->cols( int( $boundary->[ 2 ] ) - int( $boundary->[ 0 ] ) + 1 )
        ;    # pixel columns
#    print "Iterator rows|cols: ",$self->rows,"|",$self->cols,"\n";
    $self->_verify;
    return $self;
}

sub _verify {
    my ( $self ) = @_;
# my ($image,$boundary,$buffer) =
# ( $self->image, $self->boundary, $self->buffer );
# confess "Image::GeoTIFF::Tiled object required"
# unless defined $image and ref $image and $image->isa('Image::GeoTIFF::Tiled');
    my $boundary = $self->boundary;
    confess "Boundary required"
        unless defined $boundary;
    confess "Array ref of size 4 required for boundary"
        unless ref $boundary
            and ref $boundary eq 'ARRAY'
            and scalar @{ $boundary } == 4;

    my $buffer = $self->buffer;
    confess "Buffered data required"
        unless defined $buffer;

    unless (ref $buffer
        and ref $buffer eq 'ARRAY'
        and ref $buffer->[ 0 ]
        and ref $buffer->[ 0 ] eq 'ARRAY' )
    {
        confess "Buffered data needs to be a 2D arrayref";
    }
    # Note: if rows,cols = 0 then just return undef
}

sub reset {
    my $self = shift;
    $self->current_row( 0 );
    $self->current_col( -1 );
}

#================================================================================================#
# Accessors

sub _elem {
    my $self = shift;
    my $key  = shift;
    return $self->{ $key } unless @_;
    $self->{ $key } = $_[ 0 ];
}
# Config accessors
# sub image    { &_elem( shift, 'image',    @_ ); }
sub boundary { &_elem( shift, 'boundary', @_ ); }
#sub buffer { &_elem(shift,'buffer', @_); }
sub rows        { &_elem( shift, 'rows',        @_ ); }
sub cols        { &_elem( shift, 'cols',        @_ ); }
sub current_row { &_elem( shift, 'current_row', @_ ); }
sub current_col { &_elem( shift, 'current_col', @_ ); }
sub buffer      { return shift->{ buffer } }

sub get {
    my ( $self, $row, $col ) = @_;
    $self->buffer->[ $row ][ $col ];
}

sub current_coord {
    my $self     = shift;
    my $boundary = $self->{ boundary };
    # current pixel position - middle of the pixel
    my $x = $boundary->[ 0 ] + $self->current_col + 0.5;
    my $y = $boundary->[ 1 ] + $self->current_row + 0.5;
    ( $x, $y );
}

#================================================================================================#
# ITERATION METHODS

sub next {
    my ( $self ) = @_;
    my ( $rows, $cols ) = ( $self->{ rows },        $self->{ cols } );
    my ( $row,  $col )  = ( $self->{ current_row }, $self->{ current_col } );
    my $val;
    NEXT_VAL: {
        # next buffer indices
        if ( $col == $cols - 1 ) {
            $row++;
            $col = 0;
        }
        else {
            $col++;
        }

        # Check if there is a "next" coordinate
        if ( $row >= $rows
            || ( $row == $rows - 1 && $col >= $cols ) )
        {
            return;
        }
        $val = $self->buffer->[ $row ][ $col ];
        # Return the next valid value
        redo NEXT_VAL unless defined $val;
        # redo NEXT_VAL if $val == $NULL;
    }
    # Store coordinate
    $self->{ current_row } = $row;
    $self->{ current_col } = $col;

    $val;
} ## end sub next

sub adjacencies {
    my $self   = shift;
    my $buffer = $self->{ buffer };
    my ( $row, $col ) = ( $self->{ current_row }, $self->{ current_col } );
    # Return 8 adjacent values (starting NW, circling clockwise)
    my @adj = (
        ( map $buffer->[ $row - 1 ][ $_ ], $col - 1 .. $col + 1 ),
        $buffer->[ $row ][ $col + 1 ],
        ( map $buffer->[ $row + 1 ][ $_ ], reverse $col - 1 .. $col + 1 ),
        $buffer->[ $row ][ $col - 1 ]
    );
    if ( $row < 1 ) {
        # undef top row
        undef $adj[$_] for (0..2);
    }
    if ( $col < 1 ) {
        # undef left column
        undef $adj[$_] for (0,6,7);
    }
    @adj;
}

#================================================================================================#
# DUMP

sub dump_buffer {
    my ( $self ) = @_;
    print "\nBuffer:\n";
    for my $r ( 0 .. $self->rows - 1 ) {
        for my $c ( 0 .. $self->cols - 1 ) {
            print " " if $c != 0;
            my $v = $self->buffer->[ $r ][ $c ];
            # if ( $v == $NULL ) {
            if ( not defined $v ) {
                print '---';
            }
            else {
                printf( "%03i", $v );
            }
        }
        print "\n";
    }
}

#================================================================================================#
# POD
1;

=head1 NAME

Image::GeoTIFF::Tiled::Iterator - A convenience class to iterate through arbitrarily-shaped raster data.

=head1 SYNOPSIS

    use Image::GeoTIFF::Tiled;
    
    my $tiff = Image::GeoTIFF::Tiled->new( $tiff_filepath );
    my $iter = $tiff->get_iterator_pix( $px_min, $py_min, $px_max, $py_max );
    
    # Dump the buffered contents
    $iter->dump_buffer;
    
    while ( defined( my $val = $iter->next ) ) {
        printf "(%.3f,%.3f): %i\n", $tiff->pix2proj(@{$iter->current_coord}), $val;
        ... # do something based on the location, value
    }

=head1 DESCRIPTION

A convenience class to iterate through arbitrarily-shaped raster data. Returns some useful state information although there's room for additional features.

=head1 SUBROUTINES/METHODS

=head2 METHODS

=over

=item new(\%opts)

Where %opts must have the keys boundary and buffer. Used by get_iterator_* methods.

=item boundary

The boundary of the buffer.

=item buffer

Returns the reference to the data buffer (a 2D array).

=item get($row,$col)

Returns the value at row $row, column $col.

=item rows

The number of rows in the buffer.

=item cols

The number of columns in the buffer.

=item current_row

The current row corresponding to the last value passed by C<next>.

=item current_col

The current column corresponding to the last value passed by C<next>.

=item current_coord

Returns the (x,y) pixel coordinate corresponding to the last value returned by C<next> as a list. The coordinate is placed in the middle of the pixel.

=item next

Returns the next buffer value, or undef if there are no further values.

=item reset

Resets the iterator so the next() returns the first value.

=item adjacencies

Returns the 8 adjacent values to the current coordinate as a list, starting from the northwest and going clockwise.

=item dump_buffer

Pretty prints the buffer's values.

=back

=head1 COPYRIGHT & LICENSE

Copyright 2010 Blake Willmarth.

This program is free software; you can redistribute it and/or
modify it under the terms of either:

=over 4

=item * the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any
later version, or

=item * the Artistic License version 2.0.

=back

=cut
